pkg/controller/cainjector/setup.go

```diff
@@ -67,37 +77,85 @@ var (
+	controllers := map[string]controller.Controller{}
 	for _, setup := range injectorSetups {
-		if err := Register(mgr, setup, sources...); err != nil {
+		controller, err := Register(groupName, mgr, setup, sources, ca, client)
+		if err != nil {
 			if !meta.IsNoMatchError(err) || !setup.injector.IsAlpha() {
 				return err
 			}
 			ctrl.Log.V(logf.WarnLevel).Info("unable to register injector which is still in an alpha phase."+
 				" Enable the feature on the API server in order to use this injector",
 				"injector", setup.resourceName)
 		}
+		controllers[setup.resourceName] = controller
 	}
-	return nil
+	g, gctx := errgroup.WithContext(ctx)
+
+	g.Go(func() (err error) {
+		defer func() {
+			ctrl.Log.V(logf.TraceLevel).Error(err, "DEBUG: cache routine finished", "group-name", groupName)
+		}()
+		ctrl.Log.V(logf.TraceLevel).Info("DEBUG: cache routine starting", "group-name", groupName)
+		if err = ca.Start(gctx.Done()); err != nil {
+			return errors.WithStack(err)
+		}
+		return nil
+	})
+	if ca.WaitForCacheSync(gctx.Done()) {
+		for resourceName, controller := range controllers {
+			if gctx.Err() != nil {
+				continue
+			}
+			resourceName := resourceName
+			controller := controller
+			g.Go(func() (err error) {
+				defer func() {
+					ctrl.Log.V(logf.TraceLevel).Error(err, "DEBUG: controller routine finished", "group-name", groupName, "resourceName", resourceName)
+				}()
+				ctrl.Log.V(logf.TraceLevel).Info("DEBUG: controller routine starting", "group-name", groupName, "resourceName", resourceName)
+				return controller.Start(gctx.Done())
+			})
+		}
+	}
+	return g.Wait()
 }
 
 // Register registers an injection controller with the given manager, and adds relevant indicies.
-func Register(mgr ctrl.Manager, setup injectorSetup, sources ...caDataSource) error {
+func Register(groupName string, mgr ctrl.Manager, setup injectorSetup, sources []caDataSource, ca cache.Cache, client client.Client) (controller.Controller, error) {
+	log := ctrl.Log.WithName(groupName).WithName(setup.resourceName)
 	typ := setup.injector.NewTarget().AsObject()
-	builder := ctrl.NewControllerManagedBy(mgr).For(typ)
+
+	c, err := controller.NewUnmanaged(
+		// XXX:  strings.ToLower(typ.GetObjectKind().GroupVersionKind().Kind),
+		fmt.Sprintf("controller-for-%s-%s", groupName, setup.resourceName),
+		mgr,
+		controller.Options{
+			Reconciler: &genericInjectReconciler{
+				Client:       client,
+				sources:      sources,
+				log:          log.WithName("generic-inject-reconciler"),
+				resourceName: setup.resourceName,
+				injector:     setup.injector,
+			},
+			Log:                     log,
+			MaxConcurrentReconciles: 5,
+		})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if err := c.Watch(source.NewKindWithCache(typ, ca), &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, errors.WithStack(err)
+	}
```

Comment:
Setting up the controller events and cache with the different resource types becomes more long winded because we can't use the controller-runtime `controller.Builder`...instead have to plumb everything manually.